src/server/services/pipeline-in-progress/processData.js

```diff
@@ -0,0 +1,169 @@
+				startTimestamp = prevReading.endTimestamp;
+				endTimestamp = moment(rows[index - readingRepetition][1], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+			}
+			else {
+				startTimestamp = moment(rows[index - readingRepetition][1], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+				endTimestamp = moment(rows[index - readingRepetition][2], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+			}
+			if (isCumulative && isFirst(prevReading.endTimestamp)) {
+				readingOK = false;
+				errMsg = 'The first reading must be dropped when dealing with cumulative data.';
+			}
+			if (onlyEndTime && isFirst(prevReading.endTimestamp)) {
+				readingOK = false;
+				errMsg = 'The first reading must be dropped when dealing only with endTimestamps.';
+			}
+			if (readingOK && startTimestamp.isSameOrAfter(endTimestamp)) {
+				readingOK = false;
+				errMsg = 'The reading end time is not after the start time.';
+				if (onlyEndTime) {
+					errMsg += ' The start time came from the previous readings end time.';
+				}
+			}
+			if (readingOK) {
+				// Check that startTimestamp is not before the previous endTimestamp
+				if (onlyEndTime && endTimestamp.isSameOrBefore(prevReading.endTimestamp)) {
+					readingOK = false;
+					errMsg = 'The reading is not after the previous reading with only end time given so we must drop the reading.';
+				}
+				else if (startTimestamp.isBefore(prevReading.endTimestamp)) {
+					if (isCumulative) {
+						readingOK = false;
+						errMsg = 'The reading start time is before the previous endTime and cumulative so OED cannot use this reading.'
+					}
+					else if (!isFirst(prevReading.endTimestamp)) {
+						//Only treat this as a warning since the readings may be sent in a different order.
+						errMsg = 'The current reading startTime is not after the previous reading\'s end time. This is expected on the first reading ever.';
+					}
+				}
+			}
+			if (readingOK && Math.abs(startTimestamp.diff(prevReading.endTimestamp)) > Tgap) {
+				if (isCumulative) {
+					readingOK = false;
+					errMsg = 'The end of the previous reading is too far from the start of the next readings in cumulative data so drop the readings.';
+				}
+				else if (!isFirst(prevReading.endTimestamp)) {
+					//Only treat this as a warning since we expect a gap in the first ever reading.
+					errMsg = 'There is a gap in time between this reading and the previous reading.';
+				}
+			}
+			if (readingOK) {
+				//This reading can be used
+				if (isCumulative) {
+					meterReading1 = rows[index - 2 * readingRepetition][0]; // use this if regular
+					meterReading2 = rows[index - readingRepetition][0];
+					meterReading = meterReading2 - meterReading1; // use this ifCumulative
+				}
+				else {
+					meterReading = rows[index - readingRepetition][0];
+				}
+				// Reject negative readings
+				if (isCumulative && meterReading1 < 0) {
+					// If the value isCumulative and the first reading is negative reject unless cumulativeReset is true
```

Comment:
I agree it was confusing and might be difficult for future developers. I've added more comment which hopefully make it clear what each meterReading value actually represents. 

So for example in the case of cumulative,

meterReading1 // In cumulative this is the previous raw reading
meterReading2 // In cumulative this is the current raw reading
meterReading // This is the net reading value.

Also, we might want to consider changing it so that if any negative reading is detected we throw an error. I'm not sure if this is okay or not but I don't think any raw reading which is negative should ever be recorded by a meter which means it should always error.